### PR TITLE
MM-53340 - Suppress desktop notifications for `Call Started` posts, MM >= v8.1

### PR DESCRIPTION
**Pull request by cpoile**
#### Summary
- This PR uses the desktopNotificationsHook introduced in https://github.com/mattermost/mattermost/pull/23884 to supress desktop notifications for DM and GM channels for `call started` posts only.

@DHaussermann Test plan:
- The built in desktop notification for a post in a DM/GM channel should appear if:
  - Calls ringing is set to false in the system console settings
  - Calls ringing is set to true but MM is version < v8.1
  - The post is any other type other than the `call started` post (so, just like normal). In other words: this feature ONLY affects the `calls started` post types.
- The new calls desktop notification will appear if:
  - Mattermost server is >= v8.1 (you will have to manually set the version tag on the server to at least 8.1), a call is started in a DM/GM channel, calls ringing is enabled
- If the new calls desktop notification appears, the old regular message notification will NOT appear.

#### Ticket Link
- https://mattermost.atlassian.net/browse/MM-53340

### CODE REVIEW

File: webapp/src/desktop_notifications.ts [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+
+export function desktopNotificationHandler(store: Store, post: Post, channel: Channel, args: DesktopNotificationArgs): {error?: string, args?: DesktopNotificationArgs} {
+    if (args.notify) {
+        // Calls will notify if:
+        //  1. it's a custom_calls post (call has started)
+        //  2. in a DM or GM channel
+        //  3. calls is enabled and is v0.17.0+ (it is if this is running)
+        //  4. calls ringing is enabled on the server
+
+        // @ts-ignore our imported webapp types are old
```
Comment (streamer45) [inline webapp/src/desktop_notifications.ts:17]: What would it take to update them?

File: webapp/src/desktop_notifications.ts [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+
+import {ringingEnabled} from 'src/selectors';
+import {isDmGmChannel} from 'src/utils';
+
+export function desktopNotificationHandler(store: Store, post: Post, channel: Channel, args: DesktopNotificationArgs): {error?: string, args?: DesktopNotificationArgs} {
+    if (args.notify) {
+        // Calls will notify if:
+        //  1. it's a custom_calls post (call has started)
+        //  2. in a DM or GM channel
+        //  3. calls is enabled and is v0.17.0+ (it is if this is running)
```
Comment (streamer45) [inline webapp/src/desktop_notifications.ts:14]: Would the version be `>= v0.18.0`? At least for the new notification. Also I suppose there's the extra requirement of `MM >= v8.1`

File: webapp/src/index.tsx [hunk 3]
```diff
@@ -281,6 +282,10 @@ export default class Plugin {
             return slashCommandsHandler(store, joinCall, message, args);
         });
 
+        registry.registerDesktopNotificationHook?.(async (post, msgProps, channel, teamId, args) => {
+            return desktopNotificationHandler(store, post, channel, args);
+        });
```
Comment (streamer45) [inline webapp/src/index.tsx:287]: I assume it's not a problem but just confirming that this won't have any effect on unsupported MM versions.

**Review by streamer45 — APPROVED** [review]
Thanks :+1:

File: webapp/src/desktop_notifications.ts (hunk 1 above)
Comment (cpoile) [reply webapp/src/desktop_notifications.ts:17]: We'd have to rearrange webapp and export types correctly in an npm package. I've talked to Harrison about it, probably 6 months ago, and he said it was something he definitely wants to do. But given recent org changes, I'm not sure if it's on the roadmap. I could take that on, if we can fit it in between a feature?

Comment (streamer45) [reply webapp/src/desktop_notifications.ts:17]: Thanks. Not urgent, I was mostly looking to understand what kind of effort was needed and potentially get a ticket open.

File: webapp/src/desktop_notifications.ts (hunk 2 above)
Comment (cpoile) [reply webapp/src/desktop_notifications.ts]: Ah, right, thanks, fixed!

File: webapp/src/index.tsx (hunk 3 above)
Comment (cpoile) [reply webapp/src/index.tsx:287]: 👍 The `?.` prevents a crash, and nothing will happen if this is run on an unsupported version.

File: webapp/src/desktop_notifications.ts (hunk 1 above)
Comment (cpoile) [reply webapp/src/desktop_notifications.ts:18]: A ticket to estimate the effort: https://mattermost.atlassian.net/browse/MM-53667

**Comment by DHaussermann** [comment]
@cpoile I have done a round of testing on this and the business logic works as expected. 

There is one thing that may not be working as expected. There is a section about the "new calls desktop notification" I have looked for this with the ringing feature on and off and I'm not sure what you mean.
Setup:
- Created a spin-up of  v`8.1.0-rc1`
- Deployed the plugin build from this branch

When ringing feature is off I get a desktop notification as expected but it looks no different to me.
image
When ringing feature is on I don't get a desktop notification as expected and the LHS notification looks normal.

Can you clarify this for me?

**Comment by cpoile** [comment]
@DHaussermann That notification you're seeing is the default DM notification -- with ringing off, we aren't intercepting any notifications, so we should expect to see the default one (as you see). Hope that helps?

**Comment by cpoile** [comment]
@DHaussermann Not sure if I answered your question above or not, let me know, thanks!

**Comment by cpoile** [comment]
I double checked the case Dylan mentioned and it's working as designed, so I'm going to merge for the release cut today. Thanks Dylan for the thorough work!

**Pull request merged**